docs/architecture/adr-053-go-module-refactoring.md

```diff
@@ -0,0 +1,109 @@
+addition to the other complexities related to protobuf generated code that will
+be addressed in a separate ADR.
+
+Nevertheless, the desire for semantic versioning has been [strong in the
+community](https://github.com/cosmos/cosmos-sdk/discussions/10162) and the
+single go module release process has made it very hard to
+release small changes to isolated features in a timely manner. Release cycles
+often exceed six months which means small improvements done in a day or
+two get bottle-necked by everything else in the monolithic release cycle.
+
+## Decision
+
+To improve the current situation, the SDK is being refactored into multiple
+go modules within the current repository. There has been a [fair amount of
+debate](https://github.com/cosmos/cosmos-sdk/discussions/10582#discussioncomment-1813377)
+as to how to do this, with some developers arguing for larger vs smaller
+module scopes. There are pros and cons to both approaches (which will be
+discussed below in the [Consequences](#consequences) section), but the
+approach being adopted is the following:
+* a go module should generally be scoped to a specific coherent set of
+functionality (such as math, errors, store, etc.)
+* when code is removed from the core SDK and moved to a new module path, every 
+effort should be made to avoid API breaking changes in the existing code using
+aliases and wrapper types (as done in https://github.com/cosmos/cosmos-sdk/pull/10779
+and https://github.com/cosmos/cosmos-sdk/pull/11788)
+* new go modules should be moved to a standalone domain (`cosmossdk.io`) before
+being tagged as `v1.0.0` to accommodate the possibility that they may be
+better served by a standalone repository in the future
+* all go modules should follow the guidelines in https://go.dev/blog/module-compatibility
+before `v1.0.0` is tagged and should make use of `internal` packages to limit
+the exposed API surface
+* the new go module's API may deviate from the existing code where there are
+clear improvements to be made or to remove legacy dependencies (for instance on
+amino or gogo proto), as long the old package attempts
+to avoid API breakage with aliases and wrappers
+* care should be taken when simply trying to turn an existing package into a
+new go module: https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository.
+In general, it seems safer to just create a new module path (appending v2, v3, etc.
+if necessary), rather than trying to make an old package a new module.
+
+## Consequences
+
+### Backwards Compatibility
+
+If the above guidelines are followed to use aliases or wrapper types pointing
+in existing APIs that point back to the new go modules, there should be no or
+very limited breaking changes to existing APIs.
+
+### Positive
+
+* standalone pieces of software will reach `v1.0.0` sooner
+* new features to specific functionality will be released sooner 
+
+### Negative
```

Comment:
Currently SDK modules depend on other modules' public APIs. With independent go module versioning, we might need to have compatibility tables like `staking@v5` is compatible with `bank@{v3,v4}` but not with `bank@v2` etc, is that correct?

It seems this can be elegantly solved by ADR-033 and the `api/` folder, but until then, is that a concern to you too @aaronc ?